src/main/python/pgshovel/replication/streams/kafka.py

```diff
@@ -39,15 +55,26 @@ def consume(self, state):
             logger.debug('Moving to previous replication log offset: %s (current position: %s)...', offset, current)
             consumer.seek(delta, 1)
             assert consumer.offsets[0] == offset
+        else:
+            decoded = prime_for_batch_start(
+                decoded,
+                max_messages=self.prime_threshold
+            )
 
-        for offset, message in consumer:
-            decoded = self.codec.decode(message.value)
-            state = validate_state(state, offset, decoded)
+        for offset, message in decoded:
+            state = validate_state(state, offset, message)
             # XXX: This is necessary because of a bug in protocol buffer oneof.
             state = type(state).FromString(state.SerializeToString())
-            yield state, offset, decoded
+            yield state, offset, message
+
 
     @classmethod
     def configure(cls, configuration, cluster, set):
         topic = '{cluster}.{set}.mutations'.format(cluster=cluster.name, set=set)
-        return cls(cluster, set, configuration['hosts'], topic)
+        return cls(
+            cluster,
+            set,
+            configuration['hosts'],
+            topic,
+            configuration.get('prime_threshold', 100)
```

Comment:
> I worry about priming taking forever and never finishing and then the application just hanging without the user really knowing what is going on. In normal operations this shouldn't be a problem, though if there is a particularly big operation that takes until the end of the stream this could happen - though that isn't really compatible with the semantics of how the stream is supposed to work (with the loader) so it might be okay.

Yeah, I have a similar sentiment. I think at least logging what's going on will help with the visibility issue (log after every N records skipped in the `dropwhile` predicate.) I'm not sure if allowing it to run unbounded is necessarily a good default, but should at least be possible.